streams/src/test/java/org/apache/kafka/streams/state/internals/BufferValueTest.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.junit.Test;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+
+public class BufferValueTest {
+    @Test
+    public void shouldDeduplicateNullValues() {
+        final BufferValue bufferValue = new BufferValue(null, null, null, null);
+        assertSame(bufferValue.priorValue(), bufferValue.oldValue());
+    }
+
+    @Test
+    public void shouldDeduplicateIndenticalValues() {
+        final byte[] bytes = {(byte) 0};
+        final BufferValue bufferValue = new BufferValue(bytes, bytes, null, null);
+        assertSame(bufferValue.priorValue(), bufferValue.oldValue());
+    }
+
+    @Test
+    public void shouldDeduplicateEqualValues() {
+        final BufferValue bufferValue = new BufferValue(new byte[] {(byte) 0}, new byte[] {(byte) 0}, null, null);
+        assertSame(bufferValue.priorValue(), bufferValue.oldValue());
+    }
+
+    @Test
+    public void shouldStoreDifferentValues() {
```

Comment:
nit: is it worth asserting `prior` and `old` are different here and in the following two tests?  I may be overthinking, but if something changed and the values are the same, these tests still pass.

Reply:
You're right. I'll follow up with a micro-PR to add these conditions.